### PR TITLE
Optimizer: support statically initialized globals which contain pointers to other globals

### PR DESCRIPTION
**Pull request by eeckstein**
For example:
```
  var p = Point(x: 10, y: 20)
  let o = UnsafePointer(&p)
```

Also support outlined arrays with pointers to other globals. For example:
```
var g1 = 1
var g2 = 2

func f() -> [UnsafePointer<Int>] {
  return [UnsafePointer(&g1), UnsafePointer(&g2)]
}
```

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyPointerToAddress.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  ///   %3 = rel_tail_addr %2
+  ///   %4 = address_to_pointer %3
+  ///   end_borrow_%2
+  ///   // use of %4
+  ///
+  var mayBeEscapedPointerFromBorrowScope: Bool {
+    if !parentFunction.hasOwnership {
+      return false
+    }
+    switch address.accessBase {
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyPointerToAddress.swift:50]: This switch looks similar to `AccessBase::hasLocalOwnershipLifetime`, except that it's more conservative for values with none ownership

```
  case AccessBase::Box:
  case AccessBase::Class:
  case AccessBase::Tail:
    return getReference()->getOwnershipKind() != OwnershipKind::None;
```

 and

```
  case AccessBase::Unidentified:
    // Unidentified storage may be nested within object access, but this is an
    // "escaped pointer", so it is not restricted to the object's borrow scope.
    return false;
```

Would it make sense for `mayBeEscapedPointerFromBorrowScope` to call through to `hasLocalOwnershipLifetime`?  Or, if that other function isn't easily bridgeable, for a version of `hasLocalOwnershipLifetime` to be added to AccessBase in the Swift sources?

File: SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift [hunk 2]
```diff
@@ -396,3 +396,21 @@ extension Function {
   }
 }
 
+extension GlobalVariable {
+  func stripAccessInstructionFromInitializer(_ context: FunctionPassContext) {
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift:400]: Could an explanation for why it's legal to delete all the begin_access/end_access instructions be helpful?

**Review by nate-chandler — APPROVED** [review]

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Pull request merged**